Fix duplicate lsp completion requests

babyccino opened the pull request:
Sorry I accidentally closed this PR so I'm making it again. Fixes completion requests being made twice upon input when completion menu is already open. I can't see why this would be necessary

Release Notes:

- Fixes bug where lsp completion requests were being made twice

babyccino committed:
fix completion requests being duplicated when completion menu is open

babyccino committed:
Merge branch 'zed-industries:main' into lsp

babyccino committed:
fix completion requests being duplicated

SomeoneToIgnore reviewed (changes requested):
Thank you for finding this.
Two notes:

1. I'm not sure that `did_trigger_completions` as an extra field in the editor is a good solution to that though — in our case, we query the duplicate completion wrongly only in the case "editor got an edit with the completions menu opened".

Such duplication happens only in

https://github.com/zed-industries/zed/blob/f2e7c635ac367e1306df425a46d90401dbe5c573/crates/editor/src/editor.rs#L2844

just a few lines before the "proper" completion trigger:

https://github.com/zed-industries/zed/blob/f2e7c635ac367e1306df425a46d90401dbe5c573/crates/editor/src/editor.rs#L2859-L2861

So what we can do instead is to call a "special" method in this context, that would do the very same as `change_selections` does but without the extra completion triggering logic (that is being called `self.selections_did_change`).
I would've had a shared function with an extra parameter, `pub fn change_selections` that calls it with one parameter, and our place that calls it with another.

2. Let's add a test for that — this bug is subtle and we'd better ensure it does not degrade over time.
`editor_tests.rs` in the `editor` crate have examples of setting up a fake language server, we can do that and add a counter into it, open the completions menu and ensure that completion requests are not re-queried too much.

Let me know if you need help with either, happy to work on this more and find a good fix.

babyccino commented:
> Thank you for finding this. Two notes:
> 
> 1. I'm not sure that `did_trigger_completions` as an extra field in the editor is a good solution to that though — in our case, we query the duplicate completion wrongly only in the case "editor got an edit with the completions menu opened".
> 
> Such duplication happens only in
> 
> https://github.com/zed-industries/zed/blob/f2e7c635ac367e1306df425a46d90401dbe5c573/crates/editor/src/editor.rs#L2844
> 
> just a few lines before the "proper" completion trigger:
> 
> https://github.com/zed-industries/zed/blob/f2e7c635ac367e1306df425a46d90401dbe5c573/crates/editor/src/editor.rs#L2859-L2861
> 
> So what we can do instead is to call a "special" method in this context, that would do the very same as `change_selections` does but without the extra completion triggering logic (that is being called `self.selections_did_change`). I would've had a shared function with an extra parameter, `pub fn change_selections` that calls it with one parameter, and our place that calls it with another.
> 
> 2. Let's add a test for that — this bug is subtle and we'd better ensure it does not degrade over time.
>    `editor_tests.rs` in the `editor` crate have examples of setting up a fake language server, we can do that and add a counter into it, open the completions menu and ensure that completion requests are not re-queried too much.
> 
> Let me know if you need help with either, happy to work on this more and find a good fix.

Yeah I didn't really like doing it with an added parameter but I just wanted to change as little as possible about the original implementation in case there were any complex interactions. I'm a bit confused on your solution though. Is what I just pushed approximately what you were thinking?

SomeoneToIgnore reviewed (commented):
Thank you, looks much better now.

> to change as little as possible

Funny, but now the diff is less than before, with the parameter

> complex interactions

Those usually grow faster from a bloated state, when a struct has a lot of parameters which interact with each other around the methods.
Now though, we just have two tracks: for selecting with completion re-querying and without, no extra state added.

> approximately what you were thinking

Yes, looks very close to!
I would hide the whole `// -- change_selections without triggering show_completions` section into a method.
Something like

```rust

// existing public API, used in many places, we keep its signature, but change the body
pub fn change_selections<R>(
        &mut self,
        autoscroll: Option<Autoscroll>,
        cx: &mut ViewContext<Self>,
        change: impl FnOnce(&mut MutableSelectionsCollection<'_>) -> R,
    ) -> R {
    // old body is moved into that inner function and adjusted a bit
    self.change_selections_inner(autoscroll, true, cx, change)
}

pub fn change_selections_inner<R>(
        &mut self,
        autoscroll: Option<Autoscroll>,
        request_completions: bool,
        cx: &mut ViewContext<Self>,
        change: impl FnOnce(&mut MutableSelectionsCollection<'_>) -> R,
    ) -> R {
    // here goes old `change_selections` logic + new parameter handling
}

// somewhere in the handle_edit place, we use self.change_selections_inner(autoscroll, false, cx, change)
```

But that's style nits (albeit will make the code more readable) — what's more important is to combine a new test for that.
We would need to create a fake language server and assign a message handler on it, like this: https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/editor_tests.rs#L5804-L5805

We can return anything hardcoded here, as we do not care to apply anything in this test.
One caveat here for our case would be not to .await the request handling before the edit is made, otherwise nothing will send the LSP request and .await'inig will not work.

other tests around use `cx.handle_request` that handles the request once, we instead, want to have a test request counter, so `cx` method is not good for us:
https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/inlay_hint_cache.rs#L1435

when we have both, we can trigger completions and check for completion menu to appear similar to
https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/editor_tests.rs#L8425

test.

Uniting that together into a new test, we can check that on N `cx.simulate_keystroke` with the context menu open, we send N completion requests exactly.

babyccino commented:
> Thank you, looks much better now.
> 
> > to change as little as possible
> 
> Funny, but now the diff is less than before, with the parameter
> 
> > complex interactions
> 
> Those usually grow faster from a bloated state, when a struct has a lot of parameters which interact with each other around the methods. Now though, we just have two tracks: for selecting with completion re-querying and without, no extra state added.
> 
> > approximately what you were thinking
> 
> Yes, looks very close to! I would hide the whole `// -- change_selections without triggering show_completions` section into a method. Something like
> 
> ```rust
> // existing public API, used in many places, we keep its signature, but change the body
> pub fn change_selections<R>(
>         &mut self,
>         autoscroll: Option<Autoscroll>,
>         cx: &mut ViewContext<Self>,
>         change: impl FnOnce(&mut MutableSelectionsCollection<'_>) -> R,
>     ) -> R {
>     // old body is moved into that inner function and adjusted a bit
>     self.change_selections_inner(autoscroll, true, cx, change)
> }
> 
> pub fn change_selections_inner<R>(
>         &mut self,
>         autoscroll: Option<Autoscroll>,
>         request_completions: bool,
>         cx: &mut ViewContext<Self>,
>         change: impl FnOnce(&mut MutableSelectionsCollection<'_>) -> R,
>     ) -> R {
>     // here goes old `change_selections` logic + new parameter handling
> }
> 
> // somewhere in the handle_edit place, we use self.change_selections_inner(autoscroll, false, cx, change)
> ```
> 
> But that's style nits (albeit will make the code more readable) — what's more important is to combine a new test for that. We would need to create a fake language server and assign a message handler on it, like this:
> 
> https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/editor_tests.rs#L5804-L5805
> 
> We can return anything hardcoded here, as we do not care to apply anything in this test. One caveat here for our case would be not to .await the request handling before the edit is made, otherwise nothing will send the LSP request and .await'inig will not work.
> 
> other tests around use `cx.handle_request` that handles the request once, we instead, want to have a test request counter, so `cx` method is not good for us:
> 
> https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/inlay_hint_cache.rs#L1435
> 
> when we have both, we can trigger completions and check for completion menu to appear similar to
> 
> https://github.com/zed-industries/zed/blob/b53d58e5538757e5614466bde29e48ffde6a869a/crates/editor/src/editor_tests.rs#L8425
> 
> test.
> 
> Uniting that together into a new test, we can check that on N `cx.simulate_keystroke` with the context menu open, we send N completion requests exactly.

Just added those comments to check in with you I was going to get rid of them 😅. 

I've already done the test pretty much exactly as you've said. I'll push it in a sec

babyccino committed:
add tests for duplicated completion requests

babyccino committed:
Merge remote-tracking branch 'upstream/main' into lsp

SomeoneToIgnore commented on crates/editor/src/editor_tests.rs:
```diff
@@ -11373,6 +11484,7 @@ pub fn handle_completion_request(
 
     let mut request = cx.handle_request::<lsp::request::Completion, _, _>(move |url, params, _| {
         let completions = completions.clone();
+        counter.fetch_add(1, atomic::Ordering::Acquire);
```
```suggestion
        counter.fetch_add(1, atomic::Ordering::Release);
```

It's not important in this case I guess, but semantically correct to use it either this way, or we can use `SeqCst` everywhere, as it's just tests.

SomeoneToIgnore commented on crates/editor/src/editor_tests.rs:
```diff
@@ -6687,6 +6703,100 @@ async fn test_completion(cx: &mut gpui::TestAppContext) {
+    let counter_clone = counter.clone();
+    let mut request = cx.handle_request::<lsp::request::Completion, _, _>(move |_, _, _| {
+        let task_completion_item = closure_completion_item.clone();
+        counter_clone.fetch_add(1, atomic::Ordering::Acquire);
```
```suggestion
        counter_clone.fetch_add(1, atomic::Ordering::Release);
```

see below

SomeoneToIgnore commented on crates/editor/src/editor_tests.rs:
```diff
@@ -6687,6 +6703,100 @@ async fn test_completion(cx: &mut gpui::TestAppContext) {
+    assert!(request.next().await.is_some());
+    request.close();
+    assert!(request.next().await.is_none());
+    assert_eq!(counter.load(atomic::Ordering::Acquire), 4);
```
```suggestion
    assert_eq!(counter.load(atomic::Ordering::Acquire), 4, "With the completions menu open, only one LSP request should happen per input");
```

I think the last assert is not needed as we just load the atomic again.
We can add some assert docs or just `//` to describe the scenario, for future investigators.

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -2841,7 +2844,22 @@ impl Editor {
-            this.change_selections(Some(Autoscroll::fit()), cx, |s| s.select(new_selections));
+
+            let trigger_in_words = !had_active_inline_completion;
+            let completions_triggered =
```
> let completions_triggered

Do we really need that boolean?
I think if the completions did not trigger here, we do not need to trigger anything below, in `this.selections_did_change` (because what really changes?), and if it did trigger, we're sure should not trigger anything.

So, again would propose to get that back into the `change_selections_inner` method instead, as proposed in https://github.com/zed-industries/zed/pull/11875#pullrequestreview-2061392229

SomeoneToIgnore reviewed (commented):
Nice, thank you so much for testing the other code too.
And a simpler way to test that too!

I think the logic is a bit odd still, let's fix it and correct the atomics, and we're ready to merge.

babyccino committed:
implement recommendations

babyccino committed:
Merge remote-tracking branch 'upstream/main' into lsp

SomeoneToIgnore reviewed (approved):
Nice, now we have a test (that takes most of the change), and a small production code change, no extra state.
One small note would be that 
`// old body is moved into that inner function and adjusted a bit`  comment is not really needed, but it does not matter at this point.

Thank you for doing all this!

babyccino committed:
remove comment

babyccino commented:
> Nice, now we have a test (that takes most of the change), and a small production code change, no extra state. One small note would be that `// old body is moved into that inner function and adjusted a bit` commend is not really needed, but it does not matter at this point.
> 
> Thank you for doing all this!
Oops I copied your code above and forgot to remove it 😅
